hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-nodemanager/src/main/java/org/apache/hadoop/yarn/server/nodemanager/LinuxContainerExecutor.java

```diff
@@ -451,8 +451,10 @@ public void startLocalizer(LocalizerStartContext ctx)
 
     } catch (PrivilegedOperationException e) {
       int exitCode = e.getExitCode();
-      LOG.warn("Exit code from container {} startLocalizer is : {}",
-          locId, exitCode, e);
+      LOG.error("Unrecoverable issue occurred. Marking the node as unhealthy to prevent "
+          + "further containers to get scheduled on the node and cause application failures. " +
+          "Exit code from the container " + locId + "startLocalizer is : " + exitCode, e);
+      nmContext.getNodeStatusUpdater().reportException(e);
```

Comment:
If we can execute container-executor, do we need to restart NodeManager?

Reply:
@zeekling thank you for looking into this change! Yes, when we hit an unrecoverable issue with the NM, the root cause needs to be fixed and the NM manually restarted. This way the RM will not schedule applications to the node while the issue is present. When we let the RM to place containers to the faulty NM, it can lead to application failures. E.g. by reaching maximum number of application attempts when the AM was scheduled to the same node twice.

Reply:
It would be nice if it could automatically recover when container-executor is ok.

Reply:
@zeekling  thank you for your suggestion. To keep this change simple and easy to follow, I've filed your improvement idea under: [YARN-11715](https://issues.apache.org/jira/browse/YARN-11715)
NodeManager should recover by itself once the container-executor can run program again

Reply:
ok